package.json

```diff
@@ -3,11 +3,11 @@
   "version": "1.0.0",
   "description": "Build package for MB userscripts",
   "scripts": {
-    "build": "NODE_ENV=production ts-node build/build.ts",
-    "build-dev": "ts-node build/build.ts",
+    "build": "NODE_ENV=production ts-node -P tsconfig.non-composite.json -r tsconfig-paths/register build/build.ts",
```

Comment:
Can you please add a comment why we need a separate `tsconfig.non-composite.json`, `tsconfig.composite.json` and `tsconfig.json`? I don't understand why we need three different configs here and why almost all configs extend `tsconfig.composite.json` while we use `tsconfig.non-composite.json` for the build...
And what's the purpose of the root `tsconfig.json` now, just "keeping things together"?

Reply:
`tsconfig.base.json` is a base configuration which sets up the main compiler options.
`tsconfig.composite.json` is a base configuration extending `tsconfig.base.json` to set up each concrete project as a composite thing.
`tsconfig.non-composite.json` is a glue project used for tools that don't properly support composite projects, like ESLint and `tsconfig-paths` (`ts-node` is fine I think, but it uses `tsconfig-paths`, so it also needs to use the non-composite glue project). It's equivalent to the original `tsconfig.json`.
`tsconfig.json` is the glue project used by the type checker, which references all composite, concrete projects (through some layers of indirection though).

AFAIK this is like the setup used by TypeScript and `typescript-eslint` themselves. The nomenclature is a bit iffy though, and I don't like how there's like 10 `tsconfig*.json` files scattered around the repository.

Would it perhaps be better to have a directory `configs` that aggregates all of the "base" configurations and actually gives them decent names? And then only have two root `tsconfig.json` files, one for `tsc` and one for `eslint`/`ts-node` etc? If so, should we also put the leaf `tsconfig.json`s in there, or should we keep those in the specific project?

If only `tsconfig.json` had some way to specify which types were available in which directories, then we wouldn't need this mess at all 🙄

Reply:
See newer commits.